### ISSUE TITLE
the default setting of CAN get wrong validation and testing acc

### ISSUE THREAD

**Issue by HengYuD** [RL-VIG/LibFewShot#25]
dataset: miniImageNet 5way 1shot
use the default config of config/can.yaml
the value of validation and testing Acc too too big and dont know why

log:
2021-11-09 16:24:37,061 [INFO] core.trainer: {'data_root': '/media/xaserver/DATA/xxx/LibFewShot/data/miniImageNet', 'image_size': 84, 'use_memory': False, 'augment': True, 'augment_times': 1, 'augment_times_query': 1, 'device_ids': 1, 'n_gpu': 1, 'seed': 0, 'deterministic': True, 'log_name': None, 'log_level': 'info', 'log_interval': 100, 'log_paramerter': False, 'result_root': './results', 'save_interval': 10, 'save_part': ['emb_func'], 'tag': None, 'epoch': 50, 'test_epoch': 5, 'parallel_part': ['emb_func'], 'pretrain_path': None, 'resume': False, 'way_num': 5, 'shot_num': 1, 'query_num': 15, 'test_way': 5, 'test_shot': 1, 'test_query': 15, 'episode_size': 4, 'train_episode': 100, 'test_episode': 100, 'batch_size': 128, 'optimizer': {'name': 'Adam', 'kwargs': {'lr': 0.001}, 'other': {'emb_func': 0.001}}, 'lr_scheduler': {'name': 'StepLR', 'kwargs': {'gamma': 0.5, 'step_size': 10}}, 'includes': ['headers/data.yaml', 'headers/device.yaml', 'headers/misc.yaml', 'headers/model.yaml', 'headers/optimizer.yaml', 'classifiers/CAN.yaml', 'backbones/Conv64F.yaml'], 'classifier': {'name': 'CAN', 'kwargs': {'scale_cls': 7, 'num_class': 64, 'nFeat': 64, 'HW': 10}}, 'backbone': {'name': 'Conv64F', 'kwargs': {'is_flatten': False, 'is_feature': False, 'leaky_relu': True, 'negative_slope': 0.2, 'last_pool': False}}, 'tb_scale': 1.0}
2021-11-09 16:24:37,097 [INFO] core.trainer: CAN(
  (emb_func): Conv64F(
    (layer1): Sequential(
      (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (2): LeakyReLU(negative_slope=0.2, inplace=True)
      (3): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    )
    (layer2): Sequential(
      (0): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (2): LeakyReLU(negative_slope=0.2, inplace=True)
      (3): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    )
    (layer3): Sequential(
      (0): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (2): LeakyReLU(negative_slope=0.2, inplace=True)
    )
    (layer3_maxpool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (layer4): Sequential(
      (0): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
      (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (2): LeakyReLU(negative_slope=0.2, inplace=True)
    )
    (layer4_pool): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (cam_layer): CAMLayer(
    (cam): CAM(
      (conv1): ConvBlock(
        (conv): Conv2d(100, 10, kernel_size=(1, 1), stride=(1, 1))
        (bn): BatchNorm2d(10, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Conv2d(10, 100, kernel_size=(1, 1), stride=(1, 1))
    )
    (classifier): Conv2d(64, 64, kernel_size=(1, 1), stride=(1, 1))
  )
  (loss_func): CrossEntropyLoss(
    (logsoftmax): LogSoftmax(dim=1)
  )
)
2021-11-09 16:24:37,159 [INFO] core.trainer: Trainable params in the model: 119378
2021-11-09 16:24:42,820 [INFO] core.data.dataset: load 38400 image with 64 label.
2021-11-09 16:24:42,853 [INFO] core.data.dataset: load 9600 image with 16 label.
2021-11-09 16:24:42,877 [INFO] core.data.dataset: load 12000 image with 20 label.
2021-11-09 16:24:42,885 [INFO] core.trainer: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    initial_lr: 0.001
    lr: 0.001
    weight_decay: 0

Parameter Group 1
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    initial_lr: 0.001
    lr: 0.001
    weight_decay: 0
)
2021-11-09 16:24:42,899 [INFO] core.trainer: ============ Train on the train set ============
2021-11-09 16:25:02,934 [INFO] core.trainer: Epoch-(0): [100/100]	Time 0.359 (0.801)	Calc 0.126 (0.118)	Data 0.073 (0.543)	Loss 5.020 (5.060)	Acc@1 33.333 (30.907)
2021-11-09 16:25:02,941 [INFO] core.trainer:  * Acc@1 30.907
2021-11-09 16:25:02,943 [INFO] core.trainer: ============ Validation on the val set ============
2021-11-09 16:25:22,238 [INFO] core.trainer: Epoch-(0): [100/100]	Time 0.185 (0.771)	Calc 0.033 (0.045)	Data 0.149 (0.723)	Acc@1 2375.000 (2008.000)
2021-11-09 16:25:22,252 [INFO] core.trainer:  * Acc@1 2008.000 Best acc -inf
2021-11-09 16:25:22,255 [INFO] core.trainer: ============ Testing on the test set ============
2021-11-09 16:25:43,216 [INFO] core.trainer: Epoch-(0): [100/100]	Time 0.156 (0.837)	Calc 0.037 (0.036)	Data 0.117 (0.796)	Acc@1 2225.000 (2009.000)
2021-11-09 16:25:43,221 [INFO] core.trainer:  * Acc@1 2009.000 Best acc -inf
2021-11-09 16:25:43,222 [INFO] core.trainer:  * Time: 0:01:00/0:50:00
2021-11-09 16:25:43,237 [INFO] core.trainer: ============ Train on the train set ============
2021-11-09 16:26:02,739 [INFO] core.trainer: Epoch-(1): [100/100]	Time 0.324 (0.779)	Calc 0.126 (0.091)	Data 0.096 (0.579)	Loss 4.887 (4.945)	Acc@1 36.333 (33.760)
2021-11-09 16:26:02,745 [INFO] core.trainer:  * Acc@1 33.760
2021-11-09 16:26:02,747 [INFO] core.trainer: ============ Validation on the val set ============
2021-11-09 16:26:19,315 [INFO] core.trainer: Epoch-(1): [100/100]	Time 0.129 (0.662)	Calc 0.051 (0.066)	Data 0.075 (0.594)	Acc@1 2375.000 (2306.000)
2021-11-09 16:26:19,320 [INFO] core.trainer:  * Acc@1 2306.000 Best acc 2008.000
2021-11-09 16:26:19,322 [INFO] core.trainer: ============ Testing on the test set ============
2021-11-09 16:26:37,707 [INFO] core.trainer: Epoch-(1): [100/100]	Time 0.146 (0.735)	Calc 0.066 (0.051)	Data 0.079 (0.680)	Acc@1 2175.000 (2322.000)
2021-11-09 16:26:37,712 [INFO] core.trainer:  * Acc@1 2322.000 Best acc 2009.000
2021-11-09 16:26:37,714 [INFO] core.trainer:  * Time: 0:01:54/0:47:30
2021-11-09 16:26:37,732 [INFO] core.trainer: ============ Train on the train set ============
2021-11-09 16:26:55,620 [INFO] core.trainer: Epoch-(2): [100/100]	Time 0.322 (0.715)	Calc 0.065 (0.109)	Data 0.086 (0.460)	Loss 4.910 (4.890)	Acc@1 29.000 (33.613)
2021-11-09 16:26:55,626 [INFO] core.trainer:  * Acc@1 33.613
2021-11-09 16:26:55,628 [INFO] core.trainer: ============ Validation on the val set ============
2021-11-09 16:27:15,458 [INFO] core.trainer: Epoch-(2): [100/100]	Time 0.155 (0.792)	Calc 0.041 (0.038)	Data 0.111 (0.751)	Acc@1 2025.000 (2292.000)
2021-11-09 16:27:15,466 [INFO] core.trainer:  * Acc@1 2292.000 Best acc 2306.000
2021-11-09 16:27:15,473 [INFO] core.trainer: ============ Testing on the test set ============
2021-11-09 16:27:36,929 [INFO] core.trainer: Epoch-(2): [100/100]	Time 0.113 (0.857)	Calc 0.033 (0.037)	Data 0.075 (0.816)	Acc@1 2200.000 (2457.000)
2021-11-09 16:27:36,942 [INFO] core.trainer:  * Acc@1 2457.000 Best acc 2322.000
2021-11-09 16:27:36,945 [INFO] core.trainer:  * Time: 0:02:54/0:48:20
2021-11-09 16:27:36,966 [INFO] core.trainer: ============ Train on the train set ============
2021-11-09 16:27:56,671 [INFO] core.trainer: Epoch-(3): [100/100]	Time 0.304 (0.787)	Calc 0.138 (0.097)	Data 0.078 (0.574)	Loss 4.713 (4.866)	Acc@1 37.667 (34.227)
2021-11-09 16:27:56,676 [INFO] core.trainer:  * Acc@1 34.227
2021-11-09 16:27:56,678 [INFO] core.trainer: ============ Validation on the val set ============
2021-11-09 16:28:13,848 [INFO] core.trainer: Epoch-(3): [100/100]	Time 0.168 (0.686)	Calc 0.080 (0.060)	Data 0.086 (0.623)	Acc@1 2025.000 (2376.000)
2021-11-09 16:28:13,854 [INFO] core.trainer:  * Acc@1 2376.000 Best acc 2306.000
2021-11-09 16:28:13,857 [INFO] core.trainer: ============ Testing on the test set ============
2021-11-09 16:28:31,461 [INFO] core.trainer: Epoch-(3): [100/100]	Time 0.133 (0.703)	Calc 0.058 (0.053)	Data 0.071 (0.648)	Acc@1 2075.000 (2496.000)
2021-11-09 16:28:31,469 [INFO] core.trainer:  * Acc@1 2496.000 Best acc 2322.000
2021-11-09 16:28:31,472 [INFO] core.trainer:  * Time: 0:03:48/0:47:30
2021-11-09 16:28:31,494 [INFO] core.trainer: ============ Train on the train set ============
2021-11-09 16:28:50,966 [INFO] core.trainer: Epoch-(4): [100/100]	Time 0.314 (0.778)	Calc 0.120 (0.101)	Data 0.057 (0.550)	Loss 4.886 (4.818)	Acc@1 30.667 (34.667)
2021-11-09 16:28:50,971 [INFO] core.trainer:  * Acc@1 34.667
2021-11-09 16:28:50,973 [INFO] core.trainer: ============ Validation on the val set ============
2021-11-09 16:29:09,167 [INFO] core.trainer: Epoch-(4): [100/100]	Time 0.207 (0.727)	Calc 0.035 (0.048)	Data 0.160 (0.675)	Acc@1 1950.000 (2253.000)
2021-11-09 16:29:09,183 [INFO] core.trainer:  * Acc@1 2253.000 Best acc 2376.000
2021-11-09 16:29:09,195 [INFO] core.trainer: ============ Testing on the test set ============
2021-11-09 16:29:29,702 [INFO] core.trainer: Epoch-(4): [100/100]	Time 0.119 (0.819)	Calc 0.032 (0.036)	Data 0.085 (0.778)	Acc@1 2450.000 (2397.000)
2021-11-09 16:29:29,706 [INFO] core.trainer:  * Acc@1 2397.000 Best acc 2496.000
2021-11-09 16:29:29,708 [INFO] core.trainer:  * Time: 0:04:46/0:47:40
2021-11-09 16:29:29,718 [INFO] core.trainer: ============ Train on the train set ============
2021-11-09 16:29:53,272 [INFO] core.trainer: Epoch-(5): [100/100]	Time 0.389 (0.941)	Calc 0.143 (0.096)	Data 0.093 (0.760)	Loss 4.827 (4.791)	Acc@1 32.000 (35.253)
2021-11-09 16:29:53,277 [INFO] core.trainer:  * Acc@1 35.253
2021-11-09 16:29:53,279 [INFO] core.trainer: ============ Validation on the val set ============
2021-11-09 16:30:08,510 [INFO] core.trainer: Epoch-(5): [100/100]	Time 0.136 (0.608)	Calc 0.052 (0.063)	Data 0.081 (0.543)	Acc@1 2825.000 (2448.000)
2021-11-09 16:30:08,514 [INFO] core.trainer:  * Acc@1 2448.000 Best acc 2376.000
2021-11-09 16:30:08,516 [INFO] core.trainer: ============ Testing on the test set ============
2021-11-09 16:30:22,733 [INFO] core.trainer: Epoch-(5): [100/100]	Time 0.154 (0.568)	Calc 0.033 (0.060)	Data 0.119 (0.505)	Acc@1 2300.000 (2481.000)
2021-11-09 16:30:22,739 [INFO] core.trainer:  * Acc@1 2481.000 Best acc 2496.000
2021-11-09 16:30:22,741 [INFO] core.trainer:  * Time: 0:05:39/0:47:05

**Comment by wZuck** [comment]
This is caused by a bug. Please refer to #24.

Notice that configs in the `config` folder are just for testing the method's implementation. We are gradually adding configs into `config/reproduce`.

For reproducing `CAN`, you may need to check its official codes for the hyper-parameters, or you can use the following config template, which is widely used in our experiments in Table.2:
```
test_episode: 1000
train_episode: 10000
episode_size: 1
epoch: 100
image_size: 84

lr_scheduler:
  kwargs:
    T_max: 100
    eta_min: 0
  name: CosineAnnealingLR

optimizer:
  kwargs:
    lr: 0.1
    momentum: 0.9
    nesterov: true
    weight_decay: 0.0005
  name: SGD
  other:
    cam_layer: 0.1
    model_func: 0.1
```

**Comment by wZuck** [comment]
Commit d844156f058f68a93b354b44bd6828d48452eed2 has already fixed the bugs, you can re-pull the `main` branch to use the bug-free codes.

**Comment by HengYuD** [comment]
nice job
